File: packages/fleet-agent/tokenv1_test.go
package main

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/json"
	"strings"
	"testing"
)

func TestHubOrigin(t *testing.T) {
	if got := hubOrigin("wss://Fleet.Ginfo.CC/v1/device"); got != "https://fleet.ginfo.cc" {
		t.Fatalf("got %q", got)
	}
	if got := hubOrigin("fleet.ginfo.cc"); got != "https://fleet.ginfo.cc" {
		t.Fatalf("got %q", got)
	}
}

func TestLegacyFlt(t *testing.T) {
	if !isLegacyFlt("flt_" + strings.Repeat("ab", 32)) {
		t.Fatal("expected legacy")
	}
	if isTokenV1("flt_" + strings.Repeat("ab", 32)) {
		t.Fatal("legacy is not v1")
	}
	if !isTokenV1("flt_1.aaa.bbb") {
		t.Fatal("expected v1")
	}
}

func TestVerifyTokenV1AndWrap(t *testing.T) {
	priv, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatal(err)
	}
	pubDER, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		t.Fatal(err)
	}
	pubB64 := b64urlEncode(pubDER)
	claims := tokenV1Claims{
		V:   1,
		Aud: "https://fleet.ginfo.cc",
		Kid: "11111111-2222-4333-8444-555555555555",
		Pub: pubB64,
		Iat: 1,
		Sec: strings.Repeat("ab", 32),
	}
	payload, err := json.Marshal(claims)
	if err != nil {
		t.Fatal(err)
	}
	sum := sha256.Sum256(payload)
	sig, err := rsa.SignPSS(rand.Reader, priv, crypto.SHA256, sum[:], &rsa.PSSOptions{SaltLength: pssSaltLen})
	if err != nil {
		t.Fatal(err)
	}
	raw := tokenV1Prefix + b64urlEncode(payload) + "." + b64urlEncode(sig)
	got, err := verifyTokenV1(raw)
	if err != nil {
		t.Fatal(err)
	}
	if got.Kid != claims.Kid || got.Sec != claims.Sec || got.Aud != claims.Aud {
		t.Fatalf("claims %+v", got)
	}
	if isLegacyFlt(raw) {
		t.Fatal("v1 token must not be legacy")
	}

	wrap, err := wrapAuth(&priv.PublicKey, claims.Sec, strings.Repeat("cc", 32))
	if err != nil {
		t.Fatal(err)
	}
	ct, err := b64urlDecode(wrap)
	if err != nil {
		t.Fatal(err)
	}
	pt, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, priv, ct, nil)
	if err != nil {
		t.Fatal(err)
	}
	var opened wrapBody
	if err := json.Unmarshal(pt, &opened); err != nil {
		t.Fatal(err)
	}
	if opened.Sec != claims.Sec || opened.Nonce != strings.Repeat("cc", 32) {
		t.Fatalf("opened %+v", opened)
	}

	msg := []byte("v1|" + claims.Aud + "|" + claims.Kid + "|nonce")
	chSum := sha256.Sum256(msg)
	chSig, err := rsa.SignPSS(rand.Reader, priv, crypto.SHA256, chSum[:], &rsa.PSSOptions{SaltLength: pssSaltLen})
	if err != nil {
		t.Fatal(err)
	}
	if !verifyChallenge(&priv.PublicKey, claims.Aud, claims.Kid, "nonce", b64urlEncode(chSig)) {
		t.Fatal("challenge sig should verify")
	}
	if verifyChallenge(&priv.PublicKey, "https://evil.example", claims.Kid, "nonce", b64urlEncode(chSig)) {
		t.Fatal("wrong aud must fail")
	}
}

func TestLegacyTokenMessage(t *testing.T) {
	if !strings.HasPrefix(highSecUpgrade, "HIGH_SEC:") {
		t.Fatalf("upgrade message must be English HIGH_SEC, got %q", highSecUpgrade)
	}
}

File: packages/fleet-worker/tokenv1.test.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import {
  CHALLENGE_MAX_LIVE,
  HIGH_SEC_KEY_MISMATCH,
  HIGH_SEC_UPGRADE,
  audMismatch,
  createChallengeBook,
  fleetOaepValue,
  hashHubToken,
  highSecAuthorization,
  hubOrigin,
  isLegacyFlt,
  isTokenV1,
  mintTokenV1,
  nextChallengeList,
  parseAuthorization,
  signChallenge,
  unwrapAuth,
  verifyChallenge,
  verifyTokenV1,
  wrapAuth,
} from "./src/tokenv1.mjs";

const AUD = "https://fleet.ginfo.cc";

test("challenge list caps live nonces and drops the oldest", () => {
  let live = [];
  const dropped = [];
  for (let i = 0; i < CHALLENGE_MAX_LIVE + 3; i++) {
    const next = nextChallengeList(live, `n${i}`);
    live = next.list;
    dropped.push(...next.dropped);
  }
  assert.equal(live.length, CHALLENGE_MAX_LIVE);
  assert.deepEqual(dropped, ["n0", "n1", "n2"]);
  assert.deepEqual(live, ["n3", "n4", "n5", "n6", "n7", "n8", "n9", "n10"]);
  const book = createChallengeBook({ max: 2 });
  book.put("kid", "a", { userId: "u" });
  book.put("kid", "b", { userId: "u" });
  book.put("kid", "c", { userId: "u" });
  assert.equal(book.take("a"), undefined);
  assert.equal(book.take("b")?.kid, "kid");
  assert.equal(book.take("c")?.kid, "kid");
});

test("payload JSON uses a stable field order", () => {
  assert.equal(
    JSON.stringify({
      v: 1,
      aud: "https://fleet.ginfo.cc",
      kid: "kid",
      pub: "pub",
      iat: 1,
      sec: "sec",
    }),
    '{"v":1,"aud":"https://fleet.ginfo.cc","kid":"kid","pub":"pub","iat":1,"sec":"sec"}',
  );
});

test("hubOrigin normalizes scheme, host case, and strips path", () => {
  assert.equal(hubOrigin("https://fleet.ginfo.cc"), AUD);
  assert.equal(hubOrigin("https://Fleet.Ginfo.CC/v1/device"), AUD);
  assert.equal(hubOrigin("wss://fleet.ginfo.cc/v1/device"), AUD);
  assert.equal(hubOrigin("fleet.ginfo.cc"), AUD);
  assert.equal(hubOrigin("http://127.0.0.1:8787/"), "http://127.0.0.1:8787");
  assert.equal(hubOrigin(""), "");
});

test("legacy flt_ hex is rejected; flt_1 is the high-sec form", () => {
  assert.equal(isLegacyFlt("flt_" + "ab".repeat(32)), true);
  assert.equal(isTokenV1("flt_" + "ab".repeat(32)), false);
  assert.equal(isTokenV1("flt_1.aaa.bbb"), true);
  assert.equal(isLegacyFlt("flt_1.aaa.bbb"), false);
});

test("parseAuthorization reads Fleet-OAEP and Bearer", () => {
  assert.deepEqual(parseAuthorization("Fleet-OAEP kid.wrap"), { kind: "oaep", kid: "kid", wrap: "wrap" });
  assert.deepEqual(parseAuthorization("Bearer flt_old"), { kind: "bearer", token: "flt_old" });
  assert.equal(parseAuthorization("").kind, "none");
  assert.equal(fleetOaepValue("k", "w"), "Fleet-OAEP k.w");
});

test("minted token verifies, wraps, and binds aud", async () => {
  const minted = await mintTokenV1({ aud: AUD });
  assert.equal(isTokenV1(minted.raw), true);
  assert.equal(minted.aud, AUD);
  assert.equal(minted.hash, await hashHubToken(minted.sec));
  assert.match(minted.prefix, /^flt_1\./);
  const claims = await verifyTokenV1(minted.raw);
  assert.equal(claims.kid, minted.kid);
  assert.equal(claims.sec, minted.sec);
  assert.equal(claims.aud, AUD);

  const nonce = "aa".repeat(32);
  const wrap = await wrapAuth({ publicSpkiB64: minted.pub, sec: minted.sec, nonce });
  const opened = await unwrapAuth({ privatePkcs8B64: minted.priv, wrapB64: wrap });
  assert.deepEqual(opened, { sec: minted.sec, nonce });

  const sig = await signChallenge({
    privatePkcs8B64: minted.priv,
    aud: AUD,
    kid: minted.kid,
    nonce,
  });
  assert.equal(
    await verifyChallenge({ publicSpkiB64: minted.pub, aud: AUD, kid: minted.kid, nonce, sig }),
    true,
  );
  assert.equal(
    await verifyChallenge({
      publicSpkiB64: minted.pub,
      aud: "https://evil.example",
      kid: minted.kid,
      nonce,
      sig,
    }),
    false,
  );
});

test("reset keypair cannot unwrap a wrap for the old token", async () => {
  const oldTok = await mintTokenV1({ aud: AUD });
  const next = await mintTokenV1({ aud: AUD });
  const wrap = await wrapAuth({
    publicSpkiB64: oldTok.pub,
    sec: oldTok.sec,
    nonce: "bb".repeat(32),
  });
  await assert.rejects(
    () => unwrapAuth({ privatePkcs8B64: next.priv, wrapB64: wrap }),
    /HIGH_SEC|OperationError|DOMException/,
  );
  const claims = await verifyTokenV1(oldTok.raw);
  assert.notEqual(claims.kid, next.kid);
});

test("highSecAuthorization challenges then returns Fleet-OAEP", async () => {
  const minted = await mintTokenV1({ aud: AUD });
  const nonce = "cc".repeat(32);
  const sig = await signChallenge({
    privatePkcs8B64: minted.priv,
    aud: AUD,
    kid: minted.kid,
    nonce,
  });
  const fetchImpl = async (url) => {
    assert.match(url, /\/v1\/challenge\?kid=/);
    assert.match(url, new RegExp(minted.kid));
    return {
      ok: true,
      json: async () => ({ nonce, kid: minted.kid, aud: AUD, sig }),
    };
  };
  const header = await highSecAuthorization(minted.raw, AUD, fetchImpl);
  assert.match(header, /^Fleet-OAEP /);
  const parsed = parseAuthorization(header);
  assert.equal(parsed.kind, "oaep");
  assert.equal(parsed.kid, minted.kid);
  const opened = await unwrapAuth({ privatePkcs8B64: minted.priv, wrapB64: parsed.wrap });
  assert.equal(opened.sec, minted.sec);
  assert.equal(opened.nonce, nonce);
});

test("legacy Bearer and aud mismatch fail in English", async () => {
  await assert.rejects(() => highSecAuthorization("flt_" + "ab".repeat(32), AUD), {
    message: HIGH_SEC_UPGRADE,
  });
  const minted = await mintTokenV1({ aud: AUD });
  await assert.rejects(() => highSecAuthorization(minted.raw, "https://evil.example"), {
    message: audMismatch(AUD, "https://evil.example"),
  });
  assert.match(HIGH_SEC_KEY_MISMATCH, /^HIGH_SEC:/);
});
